src/v/bytes/oncore.h

```diff
@@ -29,7 +29,7 @@ class oncore final {
     void verify_shard_source_location(const char* file, int linenum) const;
 
 private:
-    const shard_id_type _owner_shard;
```

Comment:
@dotnwat you added this const in a recent compile-time related change, but is it necessary? It prevents assignment but I think assignment basically DTRT at least in the case of `allocation_units` the destination will be assigned the internals which still point to the shard associated with the source so the oncore should also be assigned with the shard of the source.

Reply:
it isn't necessary, but i'm unaware of any case in which the owner shard is not the same as the shard that created the oncore object (and could make that write-once assignment). it almost seems like if it isn't const, then it implies that some code is doing manual tracking of core owner, which defeats the sort of checks we wanted in the first place.

i'll take a closer look at this PR to understand what is happening.

Reply:
@dotnwat - right, it may be true that "the owner shard is not the same as the shard that created the oncore object" but making the member `const` is implies something even stricter: "this object can never be assigned". Now I think you can argue that assignment of an `oncore` object itself doesn't make much sense since the only state it has should not change, but this also means any object that contains an `oncore` member can't be assigned (at least not without overriding the `operator=` and manually assigning all the fields except that one, ugh), and that may be an entirely valid operation.

So one way to keep similar semantics without preventing assignment operator generation could be to assert in the `oncore` assignment operator that the LHS and RHS shard ID are the same.